Avocado Job Diff [V3]

apahim opened the pull request:
v3:
- Highlight section headers with `rst` markers.
- Use `nooption` instead of `-option` for `--diff-filter`.
- Introduce `--create-reports` mode, where we suppress the embedded diff tool and create tmp files with the content to be used in external diff tools. 
- Fix issue when using job results directories to identify jobs.
- Encoding improvements.
- Resources optimizations by not creating intermediary lists with all the content.
- Basic functional tests.

v2: #1369 
- Use the paginator for the diff.
- Allow an optional destination file parameter to be passed with `--html`.
- time formatted as in human-readable plugin.
- Pass a job-results directory as a parameter instead of a SHA1.

v1: #1365 
- Use only sections that have any diff.
- Remove the title from the `legends` in html page.
- Fix the diff-filter argparse.
- `--open-html` to enable `--html`.
- Argparse help messages improved.
- Be explict regarding the time unit (secs).

adereis commented:
Working well here. Just a few really minor things:

 * It's leaving behind an empty directory in `/var/tmp/`
 * Empty line at beginning of the files generated when using `--external-tool`
 * When using `--html` and/or `--open-browser`, the stdout report is still generated. Besides being undesired IMO, the report will be open in a browser only after we reach the end of the report inside a paginator.
 * s/Html diff file/HTML diff file/
 * In the help message for `--html`, I don't think `-` should me mentioned. It was copy/pasted from the `run --html` option help, where `-` is used with the other output formats.

With the minor issues above fixed, I would consider this feature ready (as far as usability is concerned -- I haven't reviewed the code).

apahim commented:
@adereis, I updated this PR. Some comments:

 * It's leaving behind an empty directory in `/var/tmp/`

Could you expand this topic? I don't see the directory left behind.

 * Empty line at beginning of the files generated when using `--external-tool`

All sections have a '\n' before the header to separate from the previous section. We could drop the first '\n' from the results, but I'd prefer to add a header to the files. I added **Avocado Job Report** as the first line.

 * When using `--html` and/or `--open-browser`, the stdout report is still generated. Besides being undesired IMO, the report will be open in a browser only after we reach the end of the report inside a paginator.

I disabled the standard unified diff when `--html` is used.

 * s/Html diff file/HTML diff file/

Done.

 * In the help message for `--html`, I don't think `-` should me mentioned. It was copy/pasted from the `run --html` option help, where `-` is used with the other output formats.

Done.

ldoktor commented on avocado/plugins/diff.py:
```diff
@@ -0,0 +1,427 @@
+        sysinfo_dir = os.path.join(resultsdir, 'sysinfo', pre_post)
+        sysinfo = []
+        for path, subdirs, files in os.walk(sysinfo_dir):
+            for name in files:
```
The files are not sorted, making it produce diffs on the same jobs (when the dir is copied). Would you please add `sorted` here?

ldoktor replied:
...hopefully the paths are...

apahim replied:
ok

ldoktor commented on avocado/plugins/diff.py:
```diff
@@ -0,0 +1,427 @@
+        parser.add_argument('--external-tool', action='store_true',
+                            help='Instead of show the diff using Avocado '
+                            'Diff utility, create temporary files to use '
+                            'in a custom diff tool.')
```
Based on the description I'd call this `store-reports` or something like that. From an option called `--external-tool` I'd expect it to run the tool, eg:

```
avocado diff 123 321 --external-tool meld
Creating diff files /tmp/123 /tmp/321
Running meld /tmp/123 /tmp/321
```

I like both, actually :smile: But for the first version I'd probably go with `--store-reports` or similar.

apahim replied:
I will use `--create-reports` then.

apahim commented:
@ldoktor, updated with your requests.

ldoktor commented on avocado/plugins/diff.py:
```diff
@@ -0,0 +1,427 @@
+                            '(partial) ID or test results directory.')
+
+        parser.add_argument('--html', type=str,
+                            dest='html_output',
```
I wouldn't recommend using `dest` unless necessary (eg. when pairing two `store_true/store_false` arguments) as over time it becomes hard to follow. The option should always be equal to the variable....

ldoktor commented on avocado/plugins/diff.py:
```diff
@@ -0,0 +1,427 @@
+                            dest='diff_filter',
+                            type=self._validate_filters,
+                            default=['all'],
+                            help='Diff filter: (no)cmdline,(no)time,'
```
how about: comma separated filter of diff sections

ldoktor commented on avocado/plugins/diff.py:
```diff
@@ -0,0 +1,427 @@
+            if str(cmdline1) != str(cmdline2):
+                command_line_header = ['\n',
+                                       'COMMAND LINE\n',
+                                       '============\n']
```
This would be controversal, but the convention in RST is to put 1-2 empty lines before and 1 line after the heading. For section (=) heading it's usually the 2 before and 1 after and I'd prefer that.

Also at least my vim colores the headings, making it more visible. As we don't want to make "coloring books", how about to use the `log.info` (blue) color?

I mean first sections are pretty easily to distinguish, but the more sections we have the harder it gets to find them...

apahim replied:
Well, I have my doubts we should follow rst here. The report in intended to be the most diff-able possible. Investing extra lines in nice looking headers makes them hard to show up in standard diffs, needing some manual tweaks. For now, I will revert this section identifiers and just add a (blue) color to the sections titles.

apahim replied:
Actually, I'm not going to use a color in it. Blue is used for diff snips. Yellow does no work well on white backgrounds. Also, I'm removing the `===` to not appear I'm trying to use rst format. It will be: `# SECTION`, no color.

ldoktor commented on avocado/plugins/diff.py:
```diff
@@ -0,0 +1,427 @@
+                job2_results.extend(sysinfo_header_post)
+                job2_results.extend(sysinfo_post2)
+
+        if getattr(args, 'create_reports', False):
```
well you know those options are present, you can use `args.create_reports` directly... (the same on other places) (can be tweaked after the release)

ldoktor commented on avocado/plugins/diff.py:
```diff
@@ -0,0 +1,427 @@
+
+                LOG.info('HTML diff file: %s' % args.html_output)
+
+            except Exception as e:
```
the `e` is not really a nice name. Similar convention violations are all over this patch. Can you please run `pylint/pep8` and address them?

ldoktor commented on avocado/plugins/diff.py:
```diff
@@ -0,0 +1,427 @@
+            setsid = getattr(os, 'setsid', None)
+            if not setsid:
+                setsid = getattr(os, 'setpgrp', None)
+            inout = file(os.devnull, "r+")
```
opening a file using `file` is discouraged in python2.7 and forbidden in python3. Would you please use `open` insteald?

ldoktor replied:
also shouldn't we use `with open(...) as ...:` instead?

ldoktor commented:
I'm done :smile: I haven't gone through the documentation (just skimmed through it), neither the selftests, but they can be improved later. The code looks quite good, there are few places for improvement, some little issues, but overall it looks almost mergeable.
